### ISSUE TITLE
The new API function checksum checks are not adequte structs/enums

### ISSUE THREAD

**Issue by arg0d** [mozilla/uniffi-rs#1789]
Adding new fields to a struct returned by a uniffi function does not influence the function checksum. If new bindings are generated after adding the field, and library is not rebuilt, the bindings spew errors while trying to deserialize the struct from a RustBuffer and running out of data. I think the same may apply to enums.

In a worse case scenario, swaping the fields in a struct would cause wrong data to be read, because RustBuffer size is the same, just the fields are in wrong order.

**Comment by mhammond** [comment]
The checksum code is quite confusing. Each of the uniffi_bindgen types "derives" a checksum - eg, [here is a Function](https://github.com/mozilla/uniffi-rs/blob/48b1f51f6e5406cab20f502f535b30a589a5b140/uniffi_bindgen/src/interface/function.rs#L47-L48) - so the args and return type *should* be included in the checksum.

However, it's not clear this is actually used. Our [metadata definition of a function carries a checksum field](https://github.com/mozilla/uniffi-rs/blob/48b1f51f6e5406cab20f502f535b30a589a5b140/uniffi_meta/src/lib.rs#L145) and although it is an `Option<>`, best I can tell this is *always* set to `Some()` with the value created by a simple checksum of the metadata buffer used for a function [here](https://github.com/mozilla/uniffi-rs/blob/48b1f51f6e5406cab20f502f535b30a589a5b140/uniffi_meta/src/metadata.rs#L64-L87) - but the metadata buffer for a function includes the *names* of the types, but no other info about the types.

Naively, I think this is a throwback to when UDL and procmacros were a long way apart, so I'd *think* that this checksum field should be able to be omitted from *all* metadata, meaning the derived checksums from the uniffi_bindgen types should be used. But a simple attempt at doing that is failing and I'm out of time to try and work out why.

I'll try and look a little more next week, but hopefully in the meantime someone else might be able to help me understand if what I said above is wrong?

**Comment by mhammond** [comment]
Oops, the problem with the above is that procmacro exported functions never create a uniffi_bindgen Function object, and indeed, the macro exporting a function has no insight into the content of the types. So in practice it's more like that the derived checksum functions in uniffi_bindgen are what can be removed. But it's not clear how to fix this sanely either.

**Comment by bendk** [comment]
> Oops, the problem with the above is that procmacro exported functions never create a uniffi_bindgen Function object, and indeed, the macro exporting a function has no insight into the content of the types.

Yes a checksum is calculated, but it essentially only use the type name to derive itself so this is a bug.

> But it's not clear how to fix this sanely either.

Maybe we could add checksum functions for user-defined types as well.  I don't think that would be too hard, although it could be tricky getting things to work for both UDL and proc-macros.
